### DISCUSSION TITLE
x-for scoping issues

### DISCUSSION THREAD

**Discussion by foresttoney** [alpinejs/alpine#3800] [1. Help]
I am attempting to implement a quick-cart slide-over and am running into what seems to be a scoping issue. The bug presents like the JavaScript forloop closure issue.

```
const xs = Array(5);

for (i = 0; i < xs.length; i++) {
  setTimeout(() => console.log(i), 0);
}

5
5
5
5
5
``` 

Where instead I am using an `x-for` with a nested `x-data`. Something like so (abbrv):

```
<div
  x-data="QuickCart()"
  x-on:add-to-cart.window="handleOpen()">
  <template x-teleport="body">
     <div 
      x-on:keydown.window.escape="handleClose()"
      x-show="isOpen">
      <div 
        x-on:click="handleOpen()" 
        x-show="isOpen" 
        x-transition.opacity.duration.600ms>
      </div>
      ...
            <div 
              x-on:click.away="handleClose()"
              x-show="isOpen" 
              x-transition:enter-end="translate-x-0" 
              x-transition:enter-start="translate-x-full" 
              x-transition:enter="transform transition ease-in-out duration-500 sm:duration-700" 
              x-transition:leave-end="translate-x-full" 
              x-transition:leave-start="translate-x-0" 
              x-transition:leave="transform transition ease-in-out duration-500 sm:duration-700" >
             <template x-for="item in $store.Cart.cart.items">
               <div class="flex mt-8">
                 <img :src="item.image" />
              
                 <div>
                   <h3 x-text="item.title"></h3>
                   <p x-text="item.vendor"></p>
                   <p x-money.shopify="item.price"></p>
                   <div x-data="Quantity({ quantity: item.quantity })">
                     <div>
                       <button x-bind="decrement({ onDecrement: handleQuantityChange(item.variant_id) })">
                         -
                       </button>
    
                       <p x-text="quantity"></p>
    
                       <button x-bind="increment({ onIncrement: handleQuantityChange(item.variant_id) })">
                         +
                       </button>
                     </div>
  
                     <button x-on:click="handleDelete(item.variant_id)">Remove</button>
                   </div>
                </div>
              </div>
            </template>
          </div>
    ...
  </template>
</div>
```

A single item added to the cart works as expected. Then, subsequent items added have a quantity of whatever the first item in the list was (similar to the forloop example above)

```
foo - 3 +
bar - 3 +
baz - 3 +
```

I am defining custom data for `Quick Cart` and `Quantity`. The cart is implemented as a `Store` where all reads and writes to the cart are done via the store. Somewhere, I am losing scoping.

**Comment by ekwoka** [comment]
The issue here is that you aren't using a key for the `x-for` so the `quantity` values don't move with the items.

This code is a bit convoluted, but the above is your issue.

**Comment by foresttoney** [comment]
Yes - this is exactly why it wasn't working correctly. I wasn't aware that you needed to provide a `key`, but after reviewing the docs, it makes sense. I appreciate it.

**Reply by ekwoka** [reply]
Yup it isn't totally needed for all cases, but you will end up with issues, since it will lose track of which cloned element matches with with data element.

So things like nested state would not be connected properly.

**Comment by foresttoney** [comment]
A follow up question - what do you do in a scenario in which there is no unique key. For example, the Shopify Ajax Api makes the sensible* decision to, in specific scenarios, return the same product broken over several line items.

```
{
  items: [
    {
      "id": 41900603932836,
      "key": "41900603932836:15692f47e7eff04bbd16a738edb64f42",
      "title": "Foo Product",
      "price": 3895,
      "final_price": 3895,
      "quantity": 8 
    },
    {
      "id": 41900603932836,
      "key": "41900603932836:15692f47e7eff04bbd16a738edb64f42",
      "title": "Foo Product",
      "price": 3895,
      "final_price": 0,
      "quantity": 2
    },
    {
      "id": 41900603932836,
      "key": "41900603932836:15692f47e7eff04bbd16a738edb64f42",
      "title": "Foo Product",
      "price": 3895,
      "final_price": 3895,
      "quantity": 1
    }
  ]
}
```

The only relevant details is that all reasonable IDs are duplicated across items (beyond index). I tried to come up with some composite key (even hashing the entire object) but was still seeing weird tracking behavior the `x:for` directive.

**Reply by ekwoka** [reply]
Easy, the `key` there is actually unique for the line.

Even when it's the same item in every other way.

😄

**Reply by foresttoney** [reply]
I suppose I don't understand - item.key is not unique for each line. The only thing that is unique, to the extent that it is meaningful, is item.quantity. Strangely though, even when I hashed the entire object and used that as `:key`, I was still experiencing tracking issues.

**Reply by ekwoka** [reply]
I'm pretty sure shopify has the key as totally unique per line.

I have multiple Shopify stores where the cart loop is 

```html
<template x-for="item in cartItems" :key="item.key">
```

What api are you using? the normal cart API does not split a single line item into multiple rows unless they have something unique like properties or discounts, which give them new keys.

[reference](https://osoandme.com)

**Reply by foresttoney** [reply]
Most of the time `item.key` is unique. However, when cart level discounts are applied (buy four get one) this [isn't the case](https://community.shopify.com/c/technical-q-a/cart-items-with-duplicative-variant-ids/m-p/2265004/highlight/true).

**Reply by ekwoka** [reply]
Interesting. So you can preprocess the items and merge them back into one line on the key.

Or add some discount info to the key
